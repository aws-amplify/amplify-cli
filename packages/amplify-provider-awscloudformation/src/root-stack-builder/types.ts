import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import * as iam from '@aws-cdk/aws-iam';

export interface AmplifyRootStackTemplate {
  deploymentBucket?: s3.CfnBucket;
  authRole?: iam.CfnRole;
  unauthRole?: iam.CfnRole;

  addCfnParameter(props: cdk.CfnParameterProps, logicalId: string): void;
  addCfnOutput(props: cdk.CfnOutputProps, logicalId: string): void;
  addCfnMapping(props: cdk.CfnMappingProps, logicalId: string): void;
  addCfnCondition(props: cdk.CfnConditionProps, logicalId: string): void;
  addCfnResource(props: cdk.CfnResourceProps, logicalId: string): void;
}
