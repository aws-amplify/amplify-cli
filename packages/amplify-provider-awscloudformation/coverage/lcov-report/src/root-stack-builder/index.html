
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/root-stack-builder</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> src/root-stack-builder</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">76.86% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>103/134</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">54.54% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>12/22</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">62.5% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>20/32</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">76.15% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>99/130</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="index.ts"><a href="index.ts.html">index.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	</tr>

<tr>
	<td class="file medium" data-value="root-stack-builder.ts"><a href="root-stack-builder.ts.html">root-stack-builder.ts</a></td>
	<td data-value="72.97" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 72%"></div><div class="cover-empty" style="width: 28%"></div></div>
	</td>
	<td data-value="72.97" class="pct medium">72.97%</td>
	<td data-value="37" class="abs medium">27/37</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="16" class="abs medium">8/16</td>
	<td data-value="71.42" class="pct medium">71.42%</td>
	<td data-value="35" class="abs medium">25/35</td>
	</tr>

<tr>
	<td class="file high" data-value="root-stack-transform.ts"><a href="root-stack-transform.ts.html">root-stack-transform.ts</a></td>
	<td data-value="82.81" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 82%"></div><div class="cover-empty" style="width: 18%"></div></div>
	</td>
	<td data-value="82.81" class="pct high">82.81%</td>
	<td data-value="64" class="abs high">53/64</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="12" class="abs medium">8/12</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="8" class="abs medium">6/8</td>
	<td data-value="82.81" class="pct high">82.81%</td>
	<td data-value="64" class="abs high">53/64</td>
	</tr>

<tr>
	<td class="file medium" data-value="stack-synthesizer.ts"><a href="stack-synthesizer.ts.html">stack-synthesizer.ts</a></td>
	<td data-value="67.74" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 67%"></div><div class="cover-empty" style="width: 33%"></div></div>
	</td>
	<td data-value="67.74" class="pct medium">67.74%</td>
	<td data-value="31" class="abs medium">21/31</td>
	<td data-value="37.5" class="pct low">37.5%</td>
	<td data-value="8" class="abs low">3/8</td>
	<td data-value="71.42" class="pct medium">71.42%</td>
	<td data-value="7" class="abs medium">5/7</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="30" class="abs medium">20/30</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-04-03T17:51:29.627Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    