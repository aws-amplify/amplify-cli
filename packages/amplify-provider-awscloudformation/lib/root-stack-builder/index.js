"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyRootStackTransform = void 0;
var root_stack_transform_1 = require("./root-stack-transform");
Object.defineProperty(exports, "AmplifyRootStackTransform", { enumerable: true, get: function () { return root_stack_transform_1.AmplifyRootStackTransform; } });
//# sourceMappingURL=index.js.map