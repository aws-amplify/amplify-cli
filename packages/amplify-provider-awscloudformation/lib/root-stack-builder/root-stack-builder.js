"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyRootStackOutputs = exports.AmplifyRootStack = void 0;
const cdk = __importStar(require("aws-cdk-lib"));
const s3 = __importStar(require("aws-cdk-lib/aws-s3"));
const iam = __importStar(require("aws-cdk-lib/aws-iam"));
const amplify_cli_core_1 = require("amplify-cli-core");
const CFN_TEMPLATE_FORMAT_VERSION = '2010-09-09';
const ROOT_CFN_DESCRIPTION = 'Root Stack for AWS Amplify CLI';
class AmplifyRootStack extends cdk.Stack {
    constructor(scope, id, props) {
        super(scope, id, props);
        this._cfnParameterMap = new Map();
        this.generateRootStackResources = async () => {
            this.deploymentBucket = new s3.CfnBucket(this, 'DeploymentBucket', {
                bucketName: this._cfnParameterMap.get('DeploymentBucketName').valueAsString,
            });
            this.deploymentBucket.applyRemovalPolicy(cdk.RemovalPolicy.RETAIN);
            this.authRole = new iam.CfnRole(this, 'AuthRole', {
                roleName: this._cfnParameterMap.get('AuthRoleName').valueAsString,
                assumeRolePolicyDocument: {
                    Version: '2012-10-17',
                    Statement: [
                        {
                            Sid: '',
                            Effect: 'Deny',
                            Principal: {
                                Federated: 'cognito-identity.amazonaws.com',
                            },
                            Action: 'sts:AssumeRoleWithWebIdentity',
                        },
                    ],
                },
            });
            this.unauthRole = new iam.CfnRole(this, 'UnauthRole', {
                roleName: this._cfnParameterMap.get('UnauthRoleName').valueAsString,
                assumeRolePolicyDocument: {
                    Version: '2012-10-17',
                    Statement: [
                        {
                            Sid: '',
                            Effect: 'Deny',
                            Principal: {
                                Federated: 'cognito-identity.amazonaws.com',
                            },
                            Action: 'sts:AssumeRoleWithWebIdentity',
                        },
                    ],
                },
            });
        };
        this.renderCloudFormationTemplate = () => amplify_cli_core_1.JSONUtilities.stringify(this._toCloudFormation());
        this._scope = scope;
        this.templateOptions.templateFormatVersion = CFN_TEMPLATE_FORMAT_VERSION;
        this.templateOptions.description = ROOT_CFN_DESCRIPTION;
    }
    addCfnOutput(props, logicalId) {
        new cdk.CfnOutput(this, logicalId, props);
    }
    addCfnMapping(props, logicalId) {
        new cdk.CfnMapping(this, logicalId, props);
    }
    addCfnCondition(props, logicalId) {
        new cdk.CfnCondition(this, logicalId, props);
    }
    addCfnResource(props, logicalId) {
        new cdk.CfnResource(this, logicalId, props);
    }
    addCfnParameter(props, logicalId) {
        if (this._cfnParameterMap.has(logicalId)) {
            throw new amplify_cli_core_1.AmplifyError('DuplicateLogicalIdError', {
                message: `Logical Id already exists: ${logicalId}.`,
            });
        }
        this._cfnParameterMap.set(logicalId, new cdk.CfnParameter(this, logicalId, props));
    }
    getCfnParameter(logicalId) {
        if (this._cfnParameterMap.has(logicalId)) {
            return this._cfnParameterMap.get(logicalId);
        }
        throw new amplify_cli_core_1.AmplifyError('ParameterNotFoundError', {
            message: `Cfn Parameter with LogicalId ${logicalId} doesn't exist`,
        });
    }
}
exports.AmplifyRootStack = AmplifyRootStack;
class AmplifyRootStackOutputs extends cdk.Stack {
    constructor() {
        super(...arguments);
        this.renderCloudFormationTemplate = () => amplify_cli_core_1.JSONUtilities.stringify(this._toCloudFormation());
    }
    addCfnParameter() {
        throw new amplify_cli_core_1.AmplifyFault('NotImplementedFault', {
            message: 'Method not implemented.',
        });
    }
    addCfnOutput(props, logicalId) {
        new cdk.CfnOutput(this, logicalId, props);
    }
    addCfnMapping() {
        throw new amplify_cli_core_1.AmplifyFault('NotImplementedFault', {
            message: 'Method not implemented.',
        });
    }
    addCfnCondition() {
        throw new amplify_cli_core_1.AmplifyFault('NotImplementedFault', {
            message: 'Method not implemented.',
        });
    }
    addCfnResource() {
        throw new amplify_cli_core_1.AmplifyFault('NotImplementedFault', {
            message: 'Method not implemented.',
        });
    }
}
exports.AmplifyRootStackOutputs = AmplifyRootStackOutputs;
//# sourceMappingURL=root-stack-builder.js.map