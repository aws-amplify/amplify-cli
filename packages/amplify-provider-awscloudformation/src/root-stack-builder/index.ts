export { AmplifyRootStackTransform } from './root-stack-transform';
