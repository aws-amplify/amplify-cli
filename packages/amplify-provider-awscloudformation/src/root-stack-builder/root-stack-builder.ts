import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import * as iam from '@aws-cdk/aws-iam';
import { AmplifyRootStackTemplate } from '@aws-amplify/cli-extensibility-helper';
import { IStackSynthesizer, ISynthesisSession } from '@aws-cdk/core';

const CFN_TEMPLATE_FORMAT_VERSION = '2010-09-09';
const ROOT_CFN_DESCRIPTION = 'Root Stack for AWS Amplify CLI';

export type AmplifyRootStackProps = {
  synthesizer: IStackSynthesizer;
};

export class AmplifyRootStack extends cdk.Stack implements AmplifyRootStackTemplate {
  _scope: cdk.Construct;
  deploymentBucket: s3.CfnBucket;
  authRole: iam.CfnRole;
  unauthRole: iam.CfnRole;
  private _cfnParameterMap: Map<string, cdk.CfnParameter> = new Map();

  constructor(scope: cdk.Construct, id: string, props: AmplifyRootStackProps) {
    super(scope, id, props);
    this._scope = scope;
    this.templateOptions.templateFormatVersion = CFN_TEMPLATE_FORMAT_VERSION;
    this.templateOptions.description = ROOT_CFN_DESCRIPTION;
  }

  /**
   *
   * @param props :cdk.CfnOutputProps
   * @param logicalId: : lodicalId of the Resource
   */
  addCfnOutput(props: cdk.CfnOutputProps, logicalId: string): void {
    try {
      new cdk.CfnOutput(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }

  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnMapping(props: cdk.CfnMappingProps, logicalId: string): void {
    try {
      new cdk.CfnMapping(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }

  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnCondition(props: cdk.CfnConditionProps, logicalId: string): void {
    try {
      new cdk.CfnCondition(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }
  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnResource(props: cdk.CfnResourceProps, logicalId: string): void {
    try {
      new cdk.CfnResource(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }

  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnParameter(props: cdk.CfnParameterProps, logicalId: string): void {
    try {
      if (this._cfnParameterMap.has(logicalId)) {
        throw new Error('logical Id already Exists');
      }
      this._cfnParameterMap.set(logicalId, new cdk.CfnParameter(this, logicalId, props));
    } catch (error) {
      throw new Error(error);
    }
  }

  getCfnParameter(logicalId: string): cdk.CfnParameter {
    if (this._cfnParameterMap.has(logicalId)) {
      return this._cfnParameterMap.get(logicalId);
    } else {
      throw new Error(`Cfn Parameter with LogicalId ${logicalId} doesnt exist`);
    }
  }

  generateRootStackResources = async () => {
    this.deploymentBucket = new s3.CfnBucket(this, 'DeploymentBucket', {
      bucketName: this._cfnParameterMap.get('DeploymentBucketName').valueAsString,
    });

    this.deploymentBucket.applyRemovalPolicy(cdk.RemovalPolicy.RETAIN);

    this.authRole = new iam.CfnRole(this, 'AuthRole', {
      roleName: this._cfnParameterMap.get('AuthRoleName').valueAsString,
      assumeRolePolicyDocument: {
        Version: '2012-10-17',
        Statement: [
          {
            Sid: '',
            Effect: 'Deny',
            Principal: {
              Federated: 'cognito-identity.amazonaws.com',
            },
            Action: 'sts:AssumeRoleWithWebIdentity',
          },
        ],
      },
    });

    this.unauthRole = new iam.CfnRole(this, 'UnauthRole', {
      roleName: this._cfnParameterMap.get('UnauthRoleName').valueAsString,
      assumeRolePolicyDocument: {
        Version: '2012-10-17',
        Statement: [
          {
            Sid: '',
            Effect: 'Deny',
            Principal: {
              Federated: 'cognito-identity.amazonaws.com',
            },
            Action: 'sts:AssumeRoleWithWebIdentity',
          },
        ],
      },
    });
  };

  // add Function for Custom Resource in Root stack
  /**
   *
   * @param _
   * @returns
   */
  public renderCloudFormationTemplate = (_: ISynthesisSession): string => {
    return JSON.stringify(this._toCloudFormation(), undefined, 2);
  };
}

/**
 * additional class to merge CFN parameters and CFN outputs as cdk doesnt allow same logical ID of constructs in same stack
 */
export class AmplifyRootStackOutputs extends cdk.Stack implements AmplifyRootStackTemplate {
  constructor(scope: cdk.Construct, id: string, props: AmplifyRootStackProps) {
    super(scope, id, props);
  }
  deploymentBucket?: s3.CfnBucket;
  authRole?: iam.CfnRole;
  unauthRole?: iam.CfnRole;

  addCfnParameter(props: cdk.CfnParameterProps, logicalId: string): void {
    throw new Error('Method not implemented.');
  }
  addCfnOutput(props: cdk.CfnOutputProps, logicalId: string): void {
    try {
      new cdk.CfnOutput(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }
  addCfnMapping(props: cdk.CfnMappingProps, logicalId: string): void {
    throw new Error('Method not implemented.');
  }
  addCfnCondition(props: cdk.CfnConditionProps, logicalId: string): void {
    throw new Error('Method not implemented.');
  }
  addCfnResource(props: cdk.CfnResourceProps, logicalId: string): void {
    throw new Error('Method not implemented.');
  }

  public renderCloudFormationTemplate = (_: ISynthesisSession): string => {
    return JSON.stringify((this as any)._toCloudFormation(), undefined, 2);
  };
}
