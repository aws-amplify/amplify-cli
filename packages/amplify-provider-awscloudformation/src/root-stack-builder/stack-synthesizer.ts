import { ISynthesisSession, Stack, LegacyStackSynthesizer } from '@aws-cdk/core';
import { Template } from 'amplify-cli-core';
import { AmplifyRootStack, AmplifyRootStackOutputs } from './root-stack-builder';

export class RootStackSythesizer extends LegacyStackSynthesizer {
  private stacks: Map<string, Stack> = new Map();
  private static readonly stackAssets: Map<string, Template> = new Map();

  protected synthesizeStackTemplate(stack: Stack, session: ISynthesisSession): void {
    if (stack instanceof AmplifyRootStack || stack instanceof AmplifyRootStackOutputs) {
      this.addStack(stack);
      const template = stack.renderCloudFormationTemplate(session) as string;
      const templateName = stack.node.id;
      this.setStackAsset(templateName, template);
    } else {
      throw new Error('Error synthesizing the template. Expected Stack to be either instance of AmplifyRootStack');
    }
  }

  setStackAsset(templateName: string, template: string): void {
    RootStackSythesizer.stackAssets.set(templateName, JSON.parse(template));
  }

  collectStacks(): Map<string, Template> {
    return new Map(RootStackSythesizer.stackAssets.entries());
  }

  addStack(stack: Stack) {
    this.stacks.set(stack.node.id, stack);
  }

  getStack = (stackName: string): Stack => {
    if (this.stacks.has(stackName)) {
      return this.stacks.get(stackName)!;
    }
    throw new Error(`Stack ${stackName} is not created`);
  };
}
