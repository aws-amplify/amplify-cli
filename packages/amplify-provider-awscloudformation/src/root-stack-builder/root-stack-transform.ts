import { AmplifyRootStackTemplate } from '@aws-amplify/cli-extensibility-helper';
import * as cdk from '@aws-cdk/core';
import { $TSAny, $TSContext, buildOverrideDir, CFNTemplateFormat, pathManager, Template, writeCFNTemplate } from 'amplify-cli-core';
import { printer, formatter } from 'amplify-prompts';
import * as fs from 'fs-extra';
import os from 'os';
import * as path from 'path';
import * as vm from 'vm2';
import { AmplifyRootStack, AmplifyRootStackOutputs } from './root-stack-builder';
import { RootStackSythesizer } from './stack-synthesizer';

export class AmplifyRootStackTransform {
  private app: cdk.App | undefined;
  private _rootTemplateObj: AmplifyRootStack; // Props to modify Root stack data
  private _synthesizer: RootStackSythesizer;
  private _synthesizerOutputs: RootStackSythesizer;
  private _rootTemplateObjOutputs: AmplifyRootStackOutputs;
  private _resourceName: string;

  constructor(resourceName: string) {
    this._resourceName = resourceName;
    this._synthesizer = new RootStackSythesizer();
    this.app = new cdk.App();
    this._synthesizerOutputs = new RootStackSythesizer();
  }

  public async transform(context: $TSContext): Promise<Template> {
    // generate cfn Constructs and AmplifyRootStackTemplate object to get overridden
    await this.generateRootStackTemplate();

    // apply override on Amplify Object having CDK Constructs for Root Stack
    if (context.input.command !== 'init') {
      await this.applyOverride();
    }

    // generate CFN template
    const template: Template = await this.synthesizeTemplates();

    // save stack
    if (context.input.command !== 'init') {
      await this.saveBuildFiles(context, template);
    }
    return template;
  }

  private applyOverride = async () => {
    const backendDir = pathManager.getBackendDirPath();
    const overrideFilePath = path.join(backendDir, this._resourceName);
    let isBuild = false;
    try {
      isBuild = await buildOverrideDir(backendDir, overrideFilePath);
    } catch (error) {
      printer.error(`Build error : ${error.message}`);
      throw new Error(error);
    }
    // skip if packageManager or override.ts not found
    if (isBuild) {
      const overrideCode: string = await fs.readFile(path.join(overrideFilePath, 'build', 'override.js'), 'utf-8').catch(() => {
        formatter.list(['No override File Found', `To override ${this._resourceName} run amplify override auth`]);
        return '';
      });
      const sandboxNode = new vm.NodeVM({
        console: 'inherit',
        timeout: 5000,
        sandbox: {},
        require: {
          context: 'sandbox',
          builtin: ['path'],
          external: true,
        },
      });
      try {
        sandboxNode.run(overrideCode).override(this._rootTemplateObj as AmplifyRootStackTemplate);
      } catch (err: $TSAny) {
        const error = new Error(`Skipping override due to ${err}${os.EOL}`);
        printer.error(`${error}`);
        error.stack = undefined;
        throw error;
      }
    }
  };

  /**
   * Generates Root stack Template
   * @returns CFN Template
   */
  private generateRootStackTemplate = async () => {
    this._rootTemplateObj = new AmplifyRootStack(this.app, 'AmplifyRootStack', { synthesizer: this._synthesizer });

    this._rootTemplateObj.addCfnParameter(
      {
        type: 'String',
        description: 'Name of the common deployment bucket provided by the parent stack',
        default: 'DeploymentBucket',
      },
      'DeploymentBucketName',
    );

    this._rootTemplateObj.addCfnParameter(
      {
        type: 'String',
        description: 'Name of the common deployment bucket provided by the parent stack',
        default: 'AuthRoleName',
      },
      'AuthRoleName',
    );

    this._rootTemplateObj.addCfnParameter(
      {
        type: 'String',
        description: 'Name of the common deployment bucket provided by the parent stack',
        default: 'UnAuthRoleName',
      },
      'UnauthRoleName',
    );

    // Add Outputs
    this._rootTemplateObj.addCfnOutput(
      {
        description: 'CloudFormation provider root stack Region',
        value: cdk.Fn.ref('AWS::Region'),
        exportName: cdk.Fn.sub('${AWS::StackName}-Region'),
      },
      'Region',
    );

    this._rootTemplateObj.addCfnOutput(
      {
        description: 'CloudFormation provider root stack ID',
        value: cdk.Fn.ref('AWS::StackName'),
        exportName: cdk.Fn.sub('${AWS::StackName}-StackName'),
      },
      'StackName',
    );

    this._rootTemplateObj.addCfnOutput(
      {
        description: 'CloudFormation provider root stack name',
        value: cdk.Fn.ref('AWS::StackId'),
        exportName: cdk.Fn.sub('${AWS::StackName}-StackId'),
      },
      'StackId',
    );

    this._rootTemplateObj.addCfnOutput(
      {
        value: cdk.Fn.getAtt('AuthRole', 'Arn').toString(),
      },
      'AuthRoleArn',
    );

    this._rootTemplateObj.addCfnOutput(
      {
        value: cdk.Fn.getAtt('UnauthRole', 'Arn').toString(),
      },
      'UnauthRoleArn',
    );

    await this._rootTemplateObj.generateRootStackResources();

    // add another stack as above stack duplicate logical Ids
    this._rootTemplateObjOutputs = new AmplifyRootStackOutputs(this.app, 'AmplifyRootStackOutputs', {
      synthesizer: this._synthesizerOutputs,
    });
    this._rootTemplateObjOutputs.addCfnOutput(
      {
        description: 'CloudFormation provider root stack deployment bucket name',
        value: cdk.Fn.ref('DeploymentBucketName'),
        exportName: cdk.Fn.sub('${AWS::StackName}-DeploymentBucketName'),
      },
      'DeploymentBucketName',
    );

    this._rootTemplateObjOutputs.addCfnOutput(
      {
        value: cdk.Fn.ref('AuthRole'),
      },
      'AuthRoleName',
    );

    this._rootTemplateObjOutputs.addCfnOutput(
      {
        value: cdk.Fn.ref('UnauthRole'),
      },
      'UnauthRoleName',
    );
  };

  /**
   *
   * @returns return CFN templates sunthesized by app
   */
  private synthesizeTemplates = async (): Promise<Template> => {
    this.app?.synth();
    const templates = this._synthesizer.collectStacks();
    const templatesOutput = this._synthesizerOutputs.collectStacks();
    const cfnRootStack: Template = templates.get('AmplifyRootStack');
    const cfnRootStackOutputs: Template = templatesOutput.get('AmplifyRootStackOutputs');
    Object.assign(cfnRootStack.Outputs, cfnRootStackOutputs.Outputs);
    return cfnRootStack;
  };

  private saveBuildFiles = async (context: $TSContext, template: Template) => {
    const rootStackFileName = `root-cloudformation-stack.json`;
    const rootstackFilePath = path.join(pathManager.getBackendDirPath(), this._resourceName, 'build', rootStackFileName);
    // write CFN template
    await writeCFNTemplate(template, rootstackFilePath, {
      templateFormat: CFNTemplateFormat.JSON,
    });
  };

  public getRootStack(): AmplifyRootStack {
    if (this._rootTemplateObj != null) {
      return this._rootTemplateObj;
    } else {
      throw new Error(`Root Stack Template doesn't exist.`);
    }
  }
}
