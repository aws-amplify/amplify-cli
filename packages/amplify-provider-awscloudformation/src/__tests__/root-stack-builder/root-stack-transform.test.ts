import { AmplifyRootStackTransform } from '../../root-stack-builder/root-stack-transform';
import { $TSContext } from 'amplify-cli-core';

jest.mock('amplify-cli-core');

describe('Check RootStack Template', () => {
  it('Generated rootstack template during init', async () => {
    const context_stub = {
      input: {
        command: 'init',
      },
    };

    const context_stub_typed = context_stub as unknown as $TSContext;
    // CFN transform for Root stack
    const resourceName = 'awscloudformation';

    const rootTransform = new AmplifyRootStackTransform(resourceName);
    const mock_template = await rootTransform.transform(context_stub_typed);
    expect(mock_template).toMatchSnapshot();
  });
});
