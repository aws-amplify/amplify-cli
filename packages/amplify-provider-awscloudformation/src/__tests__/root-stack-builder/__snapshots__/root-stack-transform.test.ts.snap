// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check RootStack Template Generated rootstack template during init 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "Root Stack for AWS Amplify CLI",
  "Outputs": Object {
    "AuthRoleArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "AuthRole",
          "Arn",
        ],
      },
    },
    "AuthRoleName": Object {
      "Value": Object {
        "Ref": "AuthRole",
      },
    },
    "DeploymentBucketName": Object {
      "Description": "CloudFormation provider root stack deployment bucket name",
      "Export": Object {
        "Name": Object {
          "Fn::Sub": "\${AWS::StackName}-DeploymentBucketName",
        },
      },
      "Value": Object {
        "Ref": "DeploymentBucketName",
      },
    },
    "Region": Object {
      "Description": "CloudFormation provider root stack Region",
      "Export": Object {
        "Name": Object {
          "Fn::Sub": "\${AWS::StackName}-Region",
        },
      },
      "Value": Object {
        "Ref": "AWS::Region",
      },
    },
    "StackId": Object {
      "Description": "CloudFormation provider root stack name",
      "Export": Object {
        "Name": Object {
          "Fn::Sub": "\${AWS::StackName}-StackId",
        },
      },
      "Value": Object {
        "Ref": "AWS::StackId",
      },
    },
    "StackName": Object {
      "Description": "CloudFormation provider root stack ID",
      "Export": Object {
        "Name": Object {
          "Fn::Sub": "\${AWS::StackName}-StackName",
        },
      },
      "Value": Object {
        "Ref": "AWS::StackName",
      },
    },
    "UnauthRoleArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "UnauthRole",
          "Arn",
        ],
      },
    },
    "UnauthRoleName": Object {
      "Value": Object {
        "Ref": "UnauthRole",
      },
    },
  },
  "Parameters": Object {
    "AuthRoleName": Object {
      "Default": "AuthRoleName",
      "Description": "Name of the common deployment bucket provided by the parent stack",
      "Type": "String",
    },
    "DeploymentBucketName": Object {
      "Default": "DeploymentBucket",
      "Description": "Name of the common deployment bucket provided by the parent stack",
      "Type": "String",
    },
    "UnauthRoleName": Object {
      "Default": "UnAuthRoleName",
      "Description": "Name of the common deployment bucket provided by the parent stack",
      "Type": "String",
    },
  },
  "Resources": Object {
    "AuthRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Effect": "Deny",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
              "Sid": "",
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": Object {
          "Ref": "AuthRoleName",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "DeploymentBucket": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": Object {
          "Ref": "DeploymentBucketName",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "UnauthRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Effect": "Deny",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
              "Sid": "",
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": Object {
          "Ref": "UnauthRoleName",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
